### PR TITLE
Clone node

### PR DESCRIPTION
**Pull request by knoxsp**
Add the ability to clone one or more nodes within a network.

### CODE REVIEW

File: hydra_base/lib/network.py [hunk 1]
```diff
@@ -2900,6 +2900,202 @@ def clone_network(network_id,
...
+        newnode.name = name
+    else:
+        oldname = node_to_clone.name
+        num_similar_names = db.DBSession.query(Node).filter(
+            Node.network_id==node_net.id,
+            Node.name.like(f"{node_to_clone.name}%")
+        ).all()
+
+        newname = f"{oldname} ({len(num_similar_names)})"
+        newnode.name = newname
```
Comment (pmslavin) [inline hydra_base/lib/network.py:3011]: So cloning "my_node" would result in everything beginning with "my_node" counted as a similar name: e.g. "Bury_wtw" would be similar to "Bury_wtw_East", "Bury_wtw_South", etc. and the resulting number appended in 3010 would be large even though "Bury_wtw (1)" is unused. This seems likely to surprise users.

Maybe instead check
* if node to clone ends in `([0-9]+)` then increment any match to be the next free number
* if node to clone doesn't end in `([0-9]+)` then add the next free number in this format
 
Perhaps a separate function to return the "closest" suitable name?

File: hydra_base/lib/network.py [hunk 2]
```diff
@@ -2900,6 +2900,202 @@ def clone_network(network_id,
...
+
+    if new_x is not None:
+        newnode.x = new_x
+
+    if new_y is not None:
+        newnode.y = new_y
```
Comment (pmslavin) [inline hydra_base/lib/network.py:3017]: What if `new_x` and `new_y` aren't numbers?

File: tests/test_network.py [hunk 3]
```diff
@@ -1213,3 +1213,82 @@ def test_clone_network_with_scoped_attributes_to_new_project(self, client, netwo
...
+        assert len(cloned_node_2.attributes) == len(node_to_clone.attributes)
+
+        scenario = client.get_scenario(network_with_data.scenarios[0].id)
+        original_node_data = list(filter(lambda x: x.resource_attr_id in [a.id for a in node_to_clone.attributes],
+                                         scenario.resourcescenarios))
+
+        cloned_node_data = list(filter(lambda x: x.resource_attr_id in [a.id for a in cloned_node_2.attributes],
+                                       scenario.resourcescenarios))
+
+        assert len(cloned_node_data) == len(original_node_data) #has no outputs, so has one less dataset
```
Comment (pmslavin) [inline tests/test_network.py:1255]:
```suggestion
        assert len(cloned_node_data) == len(original_node_data)
```

Outputs now included as per kwarg so lengths should be the same.

File: tests/test_network.py [hunk 4]
```diff
@@ -1213,3 +1213,82 @@ def test_clone_network_with_scoped_attributes_to_new_project(self, client, netwo
...
+
+        cloned_node_3 = client.get_node(cloned_node_id_3)
+
+        assert cloned_node_3.name == name
+        assert cloned_node_3.x == x
+        assert cloned_node_3.y == y
+
+
+
+    def test_clone_nodes(self, client, network_with_data):
```
Comment (pmslavin) [inline tests/test_network.py:1273]: Some test for the "name similarity" concept might also be useful.

**Review by pmslavin — COMMENTED** [review]
Some minor comments about validating list args when cloning multiple nodes, and general queries about how the destination name is chosen.

**Review by pmslavin — COMMENTED** [review]
All clear and tests passing; minor comment about compiling regular expressions.

File: tests/test_network.py [hunk 5]
```diff
@@ -1292,3 +1292,25 @@ def test_clone_nodes(self, client, network_with_data):
...
+
+        cloned_node_ids = client.clone_nodes([node_to_clone.id])
+
+        changed_name_node = client.get_node(update_node.id)
+
+        assert changed_name_node.name ==  updated_similar_name
+
+        cloned_node = client.get_node(cloned_node_ids[0])
+
+        assert cloned_node.name == f"{node_to_clone.name} (1)"
```
[the remark this answers is not in the document]
Comment (knoxsp) [reply tests/test_network.py:1316]: Done. Good shout.

File: hydra_base/lib/network.py [hunk 6]
```diff
@@ -2969,9 +2978,58 @@ def clone_nodes(
...
+        so this would return 'Bury_wtw', as it is the closest match.
+        Additionally, "Bury_wtw (1) would be closer than "Bury_wtw (10)"
+
+
+    If node name ends in ([0-9]+) then increment any match to be the next free number
+    If node to clone doesn't end in ([0-9]+) then add the next free number in this format
+
+    """
+    #if the node to clone ends with '(number)' like 'Bury_wtw (1)'
+    pattern = r'\((\d+)\)'
```
[the remark this answers is not in the document]
Comment (knoxsp) [reply hydra_base/lib/network.py]: Done. Good shout.

**Review by pmslavin — APPROVED** [review]
Additional tests with improved coverage pass and single compiled regex now matches parenthesised digit at end of names.

**Pull request merged**